### ISSUE TITLE
Emaultor error: `INSTALL_FAILED_NO_MATCHING_ABIS`

### ISSUE THREAD

**Issue by CaffeinatedAndroid** [ReactiveCircus/android-emulator-runner#95]
The emulator is unable to install my APK due to the error: `InstallException: INSTALL_FAILED_NO_MATCHING_ABIS: Failed to extract native libraries, res=-113`. I do not see any logical reason why it is failing. Hopefully you can provide some insight.

I am using the following config in my workflow YML file:
```yml
- name: Instrumented tests (emulator)
   uses: reactivecircus/android-emulator-runner@v2
   with:
     api-level: 29
     # Runs: instrumented tests and creates coverage report for them (does not include unit tests)
     script: ./gradlew --build-cache --console=plain createDebugCoverageReport
```

Here is a snippet of the relevant stacktraces:
```sh
> Task :app:connectedDebugAndroidTest
Unable to install /Users/runner/work/space-app-android/space-app-android/app/build/outputs/apk/debug/app-debug.apk
com.android.ddmlib.InstallException: INSTALL_FAILED_NO_MATCHING_ABIS: Failed to extract native libraries, res=-113
	at com.android.ddmlib.Device.installRemotePackage(Device.java:1168)
	at com.android.ddmlib.Device.installPackage(Device.java:998)
	at com.android.ddmlib.Device.installPackage(Device.java:974)
	at com.android.ddmlib.Device.installPackage(Device.java:963)
	at com.android.build.gradle.internal.testing.ConnectedDevice.installPackage(ConnectedDevice.java:126)
	at com.android.build.gradle.internal.testing.SimpleTestRunnable.run(SimpleTestRunnable.java:134)
	at com.android.ide.common.workers.ExecutorServiceAdapter$submit$submission$1.run(ExecutorServiceAdapter.kt:68)
	at java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(ForkJoinTask.java:1386)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:175)

com.android.build.gradle.internal.testing.ConnectedDevice > runTests[test(AVD) - 10] FAILED 
	com.android.builder.testing.api.DeviceException: com.android.ddmlib.InstallException: INSTALL_FAILED_NO_MATCHING_ABIS: Failed to extract native libraries, res=-113
		at com.android.build.gradle.internal.testing.ConnectedDevice.installPackage(ConnectedDevice.java:132)
[no message defined]
java.util.concurrent.ExecutionException: java.lang.RuntimeException: java.lang.RuntimeException: com.android.builder.testing.api.DeviceException: com.android.ddmlib.InstallException: INSTALL_FAILED_NO_MATCHING_ABIS: Failed to extract native libraries, res=-113
	at java.util.concurrent.ForkJoinTask.get(ForkJoinTask.java:1006)
```

I have used this GH Action on other projects before without issue - so thanks for that 👍 

I can provide more info if necessary.

**Comment by ychescale9** [comment]
Are you using NDK or any libraries with native dependencies?

This error usually happens when your app uses a native library that's not available in the CPU architecture of the device. This action only supports the modern system images which only support `x86` and `x86_64` architectures.

Are you able to run the tests locally on an emulator? What's the CPI/ABI and Target for the system image?

**Comment by CaffeinatedAndroid** [comment]
Thanks for prompt response.

My local Emulator is x86. I am able to run tests locally through Android Studio, but the equivalent Gradle commands in the Android Studio terminal give the same errors I saw before (didn't think to check that separately as it was "working", how tedious). 

My app installs and runs on the emulator through Android Studio, but fails again when using the equivalent Gradle command: `gradlew installDebug`.

I'm not using the NDK and don't have it installed, so I guess one of my dependencies has used it? I'll have to have a dig. If it comes to it, do you know if it is possible to recompile a dependency to work with x86?

**Comment by ychescale9** [comment]
Hmm I'm surprised it works from android studio but not from the command line. Did you run `./gradlew connectedAndroidTest`?

**Comment by ychescale9** [comment]
Oh wait you can't even install the debug APK from command line.

Do you have `abiFilter` anywhere in your build.gradle?

**Comment by CaffeinatedAndroid** [comment]
Running `./gradlew connectedAndroidTest` also generates the same errors (`connectedAndroidTest` is a prerequisite of `createDebugCoverageReport`).

I do not have `abiFilter` in any of my `**/build.gradle` files.

As much as I appreciate your help (and if you think of anything, please comment), I'll close this issue. This issue is not your responsibility as it exists outside of the GitHub Action - it is some combination of my project dependencies and Gradle (but not Android Studio!?).

Thanks for nudging me in the correct direction.

**Comment by CaffeinatedAndroid** [comment]
For future fellow travellers, I have found a workaround/solution that fixed my issues. Add the following to your module's `build.gradle` to generate separate APKs for each CPU/ABI type:

```gradle
android {
    // ...
    splits {
        abi {
            enable true
            universalApk true
        }
    }
}
```
This will build for each of the default ABIs, so in most cases you shouldn't need to manually specify which ABI. I have tested this as working on: local phone (armeabi-v7a), local emulator (x86; using `./gradlew` and Android Studio), and a GitHub CI workflow using this GitHub Action (x86).
